File: backend/app/services/scoring.py
"""Free-tempo tap scoring.

The student taps at whatever speed feels natural, so absolute timing is meaningless.
Instead the first tap is anchored to the first expected note, the student's tempo is
inferred from the overall span of their taps, and each note is judged by how far (in
beats) the corresponding tap landed from where the notation says it should.
"""

from dataclasses import dataclass

# Verdict windows, in quarter-note beats of deviation from the expected onset.
ON_TIME_WINDOW = 0.25
CLOSE_WINDOW = 0.5

# Only on-time notes count toward accuracy; early/late/wrong/missed are shown to the
# student as feedback but earn no credit. Partial credit would let a student who taps
# an uneven rhythm perfectly evenly still pass, which defeats the exercise.
PASS_THRESHOLD = 0.8


@dataclass
class NoteScore:
    index: int
    expected_beat: float
    actual_beat: float | None
    deviation_beats: float | None
    verdict: str


@dataclass
class ScoreResult:
    note_results: list[NoteScore]
    accuracy: float
    passed: bool
    inferred_ms_per_beat: float | None

    @property
    def inferred_bpm(self) -> float | None:
        if not self.inferred_ms_per_beat:
            return None
        return 60000.0 / self.inferred_ms_per_beat


def _all_missed(expected_beats: list[float]) -> ScoreResult:
    notes = [
        NoteScore(index=i, expected_beat=b, actual_beat=None, deviation_beats=None, verdict="missed")
        for i, b in enumerate(expected_beats)
    ]
    return ScoreResult(note_results=notes, accuracy=0.0, passed=False, inferred_ms_per_beat=None)


def infer_ms_per_beat(expected_beats: list[float], taps_ms: list[float]) -> float | None:
    """Tempo implied by the full span of the performance: the time from the first to
    the last tap divided by the number of beats the notation says that span covers.

    A least-squares fit is deliberately NOT used here — it would pick whatever tempo
    makes the performance look best, which lets perfectly even tapping pass for an
    uneven rhythm.
    """
    n = min(len(expected_beats), len(taps_ms))
    if n < 2:
        return None
    beat_span = expected_beats[n - 1] - expected_beats[0]
    time_span = taps_ms[n - 1] - taps_ms[0]
    if beat_span <= 0 or time_span <= 0:
        return None
    return time_span / beat_span


def score_taps(expected_beats: list[float], taps_ms: list[float]) -> ScoreResult:
    """Score a list of tap timestamps (ms) against expected onset beats.

    Taps pair 1:1 with expected notes in order; unmatched trailing notes are missed
    and surplus taps are ignored.
    """
    if not expected_beats:
        return ScoreResult(note_results=[], accuracy=0.0, passed=False, inferred_ms_per_beat=None)
    if not taps_ms:
        return _all_missed(expected_beats)

    if len(expected_beats) == 1:
        note = NoteScore(
            index=0,
            expected_beat=expected_beats[0],
            actual_beat=expected_beats[0],
            deviation_beats=0.0,
            verdict="on_time",
        )
        return ScoreResult(note_results=[note], accuracy=1.0, passed=True, inferred_ms_per_beat=None)

    ms_per_beat = infer_ms_per_beat(expected_beats, taps_ms)
    b0, t0 = expected_beats[0], taps_ms[0]

    notes: list[NoteScore] = []
    for i, expected in enumerate(expected_beats):
        if i >= len(taps_ms):
            notes.append(
                NoteScore(index=i, expected_beat=expected, actual_beat=None, deviation_beats=None, verdict="missed")
            )
            continue
        if i == 0:
            notes.append(
                NoteScore(index=0, expected_beat=expected, actual_beat=expected, deviation_beats=0.0, verdict="on_time")
            )
            continue
        if ms_per_beat is None:
            notes.append(
                NoteScore(index=i, expected_beat=expected, actual_beat=None, deviation_beats=None, verdict="wrong")
            )
            continue
        actual = b0 + (taps_ms[i] - t0) / ms_per_beat
        deviation = actual - expected
        if abs(deviation) <= ON_TIME_WINDOW:
            verdict = "on_time"
        elif abs(deviation) <= CLOSE_WINDOW:
            verdict = "early" if deviation < 0 else "late"
        else:
            verdict = "wrong"
        notes.append(
            NoteScore(
                index=i,
                expected_beat=expected,
                actual_beat=round(actual, 4),
                deviation_beats=round(deviation, 4),
                verdict=verdict,
            )
        )

    accuracy = sum(1.0 for n in notes if n.verdict == "on_time") / len(notes)
    return ScoreResult(
        note_results=notes,
        accuracy=round(accuracy, 4),
        passed=accuracy >= PASS_THRESHOLD,
        inferred_ms_per_beat=ms_per_beat,
    )


def _verdict_for_deviation(deviation: float) -> str:
    if abs(deviation) <= ON_TIME_WINDOW:
        return "on_time"
    if abs(deviation) <= CLOSE_WINDOW:
        return "early" if deviation < 0 else "late"
    return "wrong"


def score_taps_strict(
    expected_beats: list[float], taps_ms: list[float], ms_per_beat: float
) -> ScoreResult:
    """Score taps against a fixed metronome grid.

    ``taps_ms`` must be relative to the downbeat that follows the count-in (beat 0 of
    the pattern). Unlike free-tempo scoring there is no anchoring and no tempo
    inference: every note, including the first, is judged by how far it landed from
    ``expected_beat * ms_per_beat``.
    """
    if not expected_beats:
        return ScoreResult(note_results=[], accuracy=0.0, passed=False, inferred_ms_per_beat=ms_per_beat)
    if not taps_ms:
        result = _all_missed(expected_beats)
        result.inferred_ms_per_beat = ms_per_beat
        return result

    notes: list[NoteScore] = []
    for i, expected in enumerate(expected_beats):
        if i >= len(taps_ms):
            notes.append(
                NoteScore(index=i, expected_beat=expected, actual_beat=None, deviation_beats=None, verdict="missed")
            )
            continue
        actual = taps_ms[i] / ms_per_beat
        deviation = actual - expected
        notes.append(
            NoteScore(
                index=i,
                expected_beat=expected,
                actual_beat=round(actual, 4),
                deviation_beats=round(deviation, 4),
                verdict=_verdict_for_deviation(deviation),
            )
        )

    accuracy = sum(1.0 for n in notes if n.verdict == "on_time") / len(notes)
    return ScoreResult(
        note_results=notes,
        accuracy=round(accuracy, 4),
        passed=accuracy >= PASS_THRESHOLD,
        inferred_ms_per_beat=ms_per_beat,
    )

File: frontend/src/lib/scoring.ts
/**
 * TypeScript port of backend/app/services/scoring.py
 * Free-tempo and strict-tempo tap scoring.
 */

import type { NoteResult, Verdict } from '../api/types'

const ON_TIME_WINDOW = 0.25
const CLOSE_WINDOW = 0.5
export const PASS_THRESHOLD = 0.8

export interface ScoreResult {
  noteResults: NoteResult[]
  accuracy: number
  passed: boolean
  inferredMsPerBeat: number | null
}

function verdict(deviation: number): Verdict {
  if (Math.abs(deviation) <= ON_TIME_WINDOW) return 'on_time'
  if (Math.abs(deviation) <= CLOSE_WINDOW) return deviation < 0 ? 'early' : 'late'
  return 'wrong'
}

function allMissed(expectedBeats: number[]): ScoreResult {
  return {
    noteResults: expectedBeats.map((b, i) => ({
      index: i,
      expected_beat: b,
      actual_beat: null,
      deviation_beats: null,
      verdict: 'missed',
    })),
    accuracy: 0,
    passed: false,
    inferredMsPerBeat: null,
  }
}

function inferMsPerBeat(expectedBeats: number[], tapsMs: number[]): number | null {
  const n = Math.min(expectedBeats.length, tapsMs.length)
  if (n < 2) return null
  const beatSpan = expectedBeats[n - 1] - expectedBeats[0]
  const timeSpan = tapsMs[n - 1] - tapsMs[0]
  if (beatSpan <= 0 || timeSpan <= 0) return null
  return timeSpan / beatSpan
}

export function scoreTapsFree(expectedBeats: number[], tapsMs: number[]): ScoreResult {
  if (expectedBeats.length === 0) {
    return { noteResults: [], accuracy: 0, passed: false, inferredMsPerBeat: null }
  }
  if (tapsMs.length === 0) return allMissed(expectedBeats)

  if (expectedBeats.length === 1) {
    return {
      noteResults: [{ index: 0, expected_beat: expectedBeats[0], actual_beat: expectedBeats[0], deviation_beats: 0, verdict: 'on_time' }],
      accuracy: 1,
      passed: true,
      inferredMsPerBeat: null,
    }
  }

  const msPerBeat = inferMsPerBeat(expectedBeats, tapsMs)
  const b0 = expectedBeats[0]
  const t0 = tapsMs[0]

  const noteResults: NoteResult[] = expectedBeats.map((expected, i) => {
    if (i >= tapsMs.length) {
      return { index: i, expected_beat: expected, actual_beat: null, deviation_beats: null, verdict: 'missed' as Verdict }
    }
    if (i === 0) {
      return { index: 0, expected_beat: expected, actual_beat: expected, deviation_beats: 0, verdict: 'on_time' as Verdict }
    }
    if (msPerBeat === null) {
      return { index: i, expected_beat: expected, actual_beat: null, deviation_beats: null, verdict: 'wrong' as Verdict }
    }
    const actual = b0 + (tapsMs[i] - t0) / msPerBeat
    const dev = actual - expected
    return {
      index: i,
      expected_beat: expected,
      actual_beat: Math.round(actual * 10000) / 10000,
      deviation_beats: Math.round(dev * 10000) / 10000,
      verdict: verdict(dev),
    }
  })

  const accuracy = Math.round((noteResults.filter((n) => n.verdict === 'on_time').length / noteResults.length) * 10000) / 10000
  return { noteResults, accuracy, passed: accuracy >= PASS_THRESHOLD, inferredMsPerBeat: msPerBeat }
}

export function scoreTapsStrict(expectedBeats: number[], tapsMs: number[], msPerBeat: number): ScoreResult {
  if (expectedBeats.length === 0) {
    return { noteResults: [], accuracy: 0, passed: false, inferredMsPerBeat: msPerBeat }
  }
  if (tapsMs.length === 0) {
    return { ...allMissed(expectedBeats), inferredMsPerBeat: msPerBeat }
  }

  const noteResults: NoteResult[] = expectedBeats.map((expected, i) => {
    if (i >= tapsMs.length) {
      return { index: i, expected_beat: expected, actual_beat: null, deviation_beats: null, verdict: 'missed' as Verdict }
    }
    const actual = tapsMs[i] / msPerBeat
    const dev = actual - expected
    return {
      index: i,
      expected_beat: expected,
      actual_beat: Math.round(actual * 10000) / 10000,
      deviation_beats: Math.round(dev * 10000) / 10000,
      verdict: verdict(dev),
    }
  })

  const accuracy = Math.round((noteResults.filter((n) => n.verdict === 'on_time').length / noteResults.length) * 10000) / 10000
  return { noteResults, accuracy, passed: accuracy >= PASS_THRESHOLD, inferredMsPerBeat: msPerBeat }
}
